### PR TITLE
Directory based state management for model analysis tools

### PR DESCRIPTION
**Pull request by gaugup**
This PR has the following changes:-

- Move state of counterfactual, explainer, causal and erroranalysis manager to directory based state management as below.
- Add directory based state manager
- Add the unit tests for directory based state manager
- Add tests for checking the directory based storage paths for each manager state.

model_analysis_state/
|----explanations/
|----error_analysis/
|----causal/
|----counterfactual/
|--------guid1/
|------------config/
|----------------config.json
|------------data/
|----------------f1.json
|----------------f2.json
|----------------f3.json
|------------explainer/
|----------------model1.pkl
|----------------model2.pkl
|--------guid2/
|------------config/
|----------------config.json
|------------data/
|----------------f1.json
|----------------f2.json
|----------------schema.json
|------------explainer/
|----------------model1.pkl
|----------------conda_env.yml
|--------guid3/
|----model.pkl
|----train.json
|----test.json

### CODE REVIEW

**Review by riedgar-ms — APPROVED** [review]
Thanks!

File: responsibleai/responsibleai/_tools/shared/state_directory_management.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        data_directory = (Path(self.parent_directory_path) /
+                          self.sub_directory_name /
+                          DirectoryManager.DATA)
+        if not data_directory.exists():
+            data_directory.mkdir(parents=True, exist_ok=True)
+
+        return data_directory
+
+    def create_explainer_directory(self):
+        """Create the explainer directory and return its path.
```
Comment (imatiach-msft) [inline responsibleai/responsibleai/_tools/shared/state_directory_management.py:72]: this seems something too specific to explanations to be in this class?  Just seems a little weird to put it in a more common RAI class.  I wonder if there is a better way to structure the DirectoryManager.

File: responsibleai/tests/test_model_analysis.py [hunk 2]
```diff
@@ -252,6 +261,36 @@ def run_model_analysis(model, train_data, test_data, target_column,
...
+        dm = DirectoryManager(path / manager_type, component_path)
+
+        config_path = dm.get_config_directory()
+        data_path = dm.get_data_directory()
+        explainer_path = dm.get_explainer_directory()
+
+        if manager_type == ManagerNames.EXPLAINER:
+            assert not config_path.exists()
+            assert data_path.exists()
+            assert not explainer_path.exists()
```
Comment (imatiach-msft) [inline responsibleai/tests/test_model_analysis.py:279]: It seems none of them are using the explainer_path right now?  Perhaps at this point it shouldn't exist then?

File: responsibleai/responsibleai/_tools/shared/state_directory_management.py (hunk 1 above)
Comment (imatiach-msft) [reply responsibleai/responsibleai/_tools/shared/state_directory_management.py:72]: Will this also be used for causal model in the future?  Maybe the name could just be made more generic somehow.

Comment (imatiach-msft) [reply responsibleai/responsibleai/_tools/shared/state_directory_management.py:72]: It looks like it's not used currently by the managers if I understand the code correctly right now.

**Review by imatiach-msft — APPROVED** [review]

File: responsibleai/responsibleai/_tools/shared/state_directory_management.py (hunk 1 above)
Comment (imatiach-msft) [reply responsibleai/responsibleai/_tools/shared/state_directory_management.py:72]: How about this: we have different directory managers for different components, and the explanation and counterfactual components can use the ExplainerDirectoryManager which includes this explainer directory.

File: responsibleai/tests/test_model_analysis.py (hunk 2 above)
Comment (gaugup) [reply responsibleai/tests/test_model_analysis.py:279]: I plan to move causal to explainer path. The causal is state management is a little bit tricky. So keeping it in data directory for now. Will address that in up coming PR.

File: responsibleai/responsibleai/_tools/shared/state_directory_management.py (hunk 1 above)
Comment (gaugup) [reply responsibleai/responsibleai/_tools/shared/state_directory_management.py:72]: As per the design dicussion we decided on maintaining the same directory structure for all the components. Do you have any name suggestion for explainer_path? 

For causal manager, I will move the causal models to explainer path in a follow up PR. The state management for causal is alittle bit tricky.

**Pull request merged**